torch/optim/_muon.py

```diff
@@ -0,0 +1,276 @@
+    # """Configuration used by :func:`zeropower_via_newtonschulz`."""
+
+    coefficients: tuple[float, float, float] = (DEFAULT_A, DEFAULT_B, DEFAULT_C)
+    eps: float = EPS
+    ns_steps: int = 5
+
+
+AdjustLrFn: TypeAlias = Callable[[float, torch.Size], float]
+MsignFn: TypeAlias = Callable[[Tensor, BaseMsignFnConfig], Tensor]
+
+
+def zeropower_via_newtonschulz(grad: Tensor, ns_config: BaseMsignFnConfig) -> Tensor:
+    """
+    Newton-Schulz iteration to compute the zeroth power / orthogonalization of G. We opt to use a
+    quintic iteration whose coefficients are selected to maximize the slope at zero. For the purpose
+    of minimizing steps, it turns out to be empirically effective to keep increasing the slope at
+    zero even beyond the point where the iteration no longer converges all the way to one everywhere
+    on the interval. This iteration therefore does not produce UV^T but rather something like US'V^T
+    where S' is diagonal with S_{ii}' ~ Uniform(0.5, 1.5), which turns out not to hurt model
+    performance at all relative to UV^T, where USV^T = G is the SVD.
+
+    Implementation reference: https://github.com/KellerJordan/Muon/blob/master/muon.py
+    with suggestions by @jxbz, @leloykun, and @YouJiacheng.
+    """
+    ns_config = cast(NewtonSchulzConfig, ns_config)
+    steps = ns_config.ns_steps
+    eps = ns_config.eps
+    coefficients = ns_config.coefficients
+    assert steps < 100, (
+        "Number of steps must be less than 100 for computational efficiency"
+    )
+    assert len(grad.shape) == 2, "Input tensor gradient must be a 2D matrix"
+    assert len(coefficients) == 3, "Coefficients must be a tuple of exactly 3 values"
+    a, b, c = coefficients[0], coefficients[1], coefficients[2]
+    X = grad.bfloat16()
+    if grad.size(0) > grad.size(1):
+        X = X.T
+    # Ensure spectral norm is at most 1
+    X = X / (X.norm() + eps)
+    # Perform the NS iterations
+    for _ in range(steps):
+        A = X @ X.T
+        B = b * A + c * A @ A
+        X = a * X + B @ X
+
+    if grad.size(0) > grad.size(1):
+        X = X.T
+    return X
+
+
+def default_adjust_lr(lr: float, param_shape: torch.Size) -> float:
+    """Default learning rate adjustment used by Muon. Method reported in the paper https://arxiv.org/pdf/2502.16982."""
+    A, B = param_shape[:2]
+    adjusted_ratio = 0.2 * math.sqrt(max(A, B))
+    return lr * adjusted_ratio
+
+
+class Muon(Optimizer):
+    """Implements the Muon optimizer.
+
+    This optimizer performs momentum SGD followed by an optional orthogonalization
+    step computed via a user provided callable.
+    """
+
+    def __init__(
+        self,
+        params: ParamsT,
+        lr: float = 1e-3,
+        weight_decay: float = 0.1,
+        momentum: float = 0.95,
+        nesterov: bool = True,
+        *,
+        msign_fn: MsignFn = zeropower_via_newtonschulz,
+        msign_fn_config: BaseMsignFnConfig = NewtonSchulzConfig(),
```

Comment:
What is the pro of having these config live in the constructor as a struct vs separate values? Is this because these values are only used if the msign_fb is zeropower_via_newtonschulz? If so, should this config not live in the Muon constructor at all but be customizable by the user input msign_fn? What are your thoughts?

Reply:
I also wonder if this config can be made in regular dict, accepted in constructor as `Muon(..., msign_fn_config = {'eps' : 1e-5})`, and then just passed as `self.msign_fn(..., **self.msign_fn_config)` - like so, it could be easier saved into a `state_dict()`...

Reply:
I think it’s better to encapsulate the configs in a dedicated class, so the function signature stays clean and manageable. Just a preference carried over from my C++ days :)

I see Vadim's point but not sure if it's feasible (or necessity) to store the callable to state_dict in the first place.

Reply:
Another option is to have config set as simple args to the function, and then have the user override them via calling functools.partial